server/gameconnection.py

```diff
@@ -536,13 +537,21 @@ def _mark_dirty(self):
             if peer == self:
                 continue
 
-            tasks.append(peer.send_DisconnectFromPeer(self.player.id))
-
-        for result in await asyncio.gather(*tasks, return_exceptions=True):
-            if isinstance(result, Exception):
-                self._logger.exception(
-                    "peer_sendDisconnectFromPeer failed for player %i",
-                    self.player.id)
+            try:
+                tasks.append(peer.send_DisconnectFromPeer(self.player.id))
+            except AttributeError:
```

Comment:
This looks very fishy, is it possible to fix lower down? If it's difficult, we can let it slide as a hotfix, but then let's add a FIXME here.

Reply:
Not sure. Basically what happens sometimes (it's pretty rare) is that a player can disconnect causing their `lobby_connection` object to be deleted. Since players only have a weak reference to the lobby connection, calls like `player.lobby_connection.send` can fail with an `AttributeError` caused by `None.send`.

Reply:
Alright, let's just put a FIXME comment here then.

Reply:
So what exactly is the fixme? I see this as the same problem as the classic "check if file exists, then do something" problem where state can change between the check happening and the logic.

```python
if player.lobby_connection:
    # Maybe the GIL saves us here, but I still see this as bad practice
    await player.lobby_connection.send("msg")
```
vs
```python
try:
    await player.lobby_connection.send("msg")
except AttributeError:
    print("Player has disconnected while we tried to contact them")
```

Reply:
I think in the party system PR, we refactored this to add a `send_message` function to player which can handle this error in one place. So instead of `player.lobby_connection.send` you would just call `player.send_message` which we could guarantee not to throw an exception.

Reply:
Catching `AttributeError` can hide genuine programmer errors. We *need* to check for `None` further down.

Reply:
There's also no TOCTOU here, since we're not awaiting between checking the weakref and using it.

Reply:
Also, I'd think, why won't we check `player.lobby_connection` for `isinstance`/`None` on place, from where attribute error is raised?
Like, yes: we can handle exception, but it's not the fix. Problem might be hidden else.

Reply:
Ok, from the docs it says that checking for None is in fact the right thing to do.
```python
# r is a weak reference object
o = r()
if o is None:
    # referent has been garbage collected
    print("Object has been deallocated; can't frobnicate.")
else:
    print("Object is still live!")
    o.do_something_useful()
```
`the idiom shown above is safe in threaded applications as well as single-threaded applications.`